examples/ex-07.c

```diff
@@ -0,0 +1,772 @@
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License (as published by the Free
+ * Software Foundation) version 2.1 dated February 1999.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the IMPLIED WARRANTY OF MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the terms and conditions of the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+ ***********************************************************************EHEADER*/
+
+/**
+ * Example:       ex-07.c
+ *
+ * Interface:     C
+ *
+ * Requires:      only C-language support
+ *
+ * Compile with:  make ex-07
+ *
+ * Help with:     ./ex-07 -help
+ *
+ * Sample run:    mpirun -np 2 ex-07 -rank 3
+ *
+ * Description:   Solve the chaotic Lorenz system,
+ *
+ *                  {x' = sigma (y - x),    x(0) = -1.8430428,
+ *                  {y' = x (rho - z) - y,  y(0) = -0.07036326,
+ *                  {z' = xy - beta z,      z(0) = 23.15614636.
+ *
+ *                with sigma=10, beta=8/3, and rho=28,
+ *                and estimate the backward Lyapunov vectors along the trajectory
+ *                parallel-in-time using MGRIT equipped with Delta correction.
+ *                The initial condition should be a point close to the strange attractor.
+ *
+ *                TODO: Give some high-level info about the algorithm.
+ *
+ *                See (https://arxiv.org/abs/2208.12629) for a detailed introduction
+ *                to the technique used here.
+ *
+ * Visualize with: python viz-ex-07.py (requires NumPy and MatPlotLib)
+ *
+ **/
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "braid.h"
+#include "braid_test.h"
+
+/*-------------------------*
+ | linear algebra routines |
+ *-------------------------*/
+
+#define VecSize 3
+typedef double VEC[VecSize];
+typedef double MAT[VecSize][VecSize];
+
+const double sigma = 10.0;
+const double beta = 8.0 / 3.0;
+const double rho = 28.0;
+
+void VecCopy(const VEC x, VEC y)
+{
+   for (int i = 0; i < VecSize; i++)
+   {
+      y[i] = x[i];
+   }
+}
+
+void VecSet(VEC x, double alpha)
+{
+   for (int i = 0; i < VecSize; i++)
+   {
+      x[i] = alpha;
+   }
+}
+
+void MatSetEye(MAT I)
+{
+   for (size_t i = 0; i < VecSize; i++)
+   {
+      VecSet(I[i], 0);
+      I[i][i] = 1.;
+   }
+}
+
+void VecAxpy(const double alpha, const VEC x, VEC y)
+{
+   for (int i = 0; i < VecSize; i++)
+   {
+      y[i] = alpha * x[i] + y[i];
+   }
+}
+
+void MatAxpy(const double alpha, const MAT A, MAT B)
+{
+   for (int i = 0; i < VecSize; i++)
+   {
+      VecAxpy(alpha, A[i], B[i]);
+   }
+}
+
+void MatVec(const MAT A, VEC x)
+{
+   VEC tmp;
```

Comment:
is Vec tmp every deallocated?

doing a valgrind on some various parameter choices, with leakcheck=full, would probably be good